### PR TITLE
[RFR] First advice

### PR DESCRIPTION
**Pull request by HALLERPierre**
[Trello](https://trello.com/c/XhLmxOa3/1-1-as-fred-given-its-my-turn-i-want-to-ask-for-the-move-that-would-make-me-win-the-game-175)

# To Do
- [x] Rewrite all game logic
- [x] Add simulation class to play all possibles moves
- [x] Add a scorer class to assign a score to a given board
- [x] Add server
- [x] Handle client request
- [x] Serialize response
- [x] Add tests

# Result
win

### CODE REVIEW

File: advisor-go/src/quixo/game/game.go [hunk 1]
```diff
@@ -1,2 +1,167 @@
...
+}
+
+// GetAvailablesDestinations return an array of destinations
+func GetAvailablesDestinations(grid [][]int, coords Coords) []Coords {
+	destinations := []Coords{}
+	x := coords.X
+	y := coords.Y
+	indexEnd := len(grid) - 1
+
+	if x == 0 || x == indexEnd {
```
Comment (fzaninotto) [inline advisor-go/src/quixo/game/game.go]: this part is hard to follow.

File: advisor-go/src/quixo/game/game.go [hunk 2]
```diff
@@ -1,2 +1,167 @@
...
+		return FlipRowsAndCols(flippedGrid)
+	}
+	return nil
+}
+
+func isOutsideCube(coords Coords, size int) bool {
+	return coords.X == 0 || coords.Y == 0 || coords.X == size-1 || coords.Y == size-1
+}
+
+func isPlayerCube(cubeValue int, playerValue int) bool {
```
Comment (fzaninotto) [inline advisor-go/src/quixo/game/game.go]: I advise you to use the same order for the function name as for the parameters: `isPlayerCube` => `cubeBelongsToPlayer`

File: advisor-go/src/quixo/game/game.go [hunk 3]
```diff
@@ -1,2 +1,167 @@
...
+	return cubeValue == 0 || cubeValue == playerValue
+}
+
+// Shift all value from xEnd to xStart in one row
```
Comment (fzaninotto) [inline advisor-go/src/quixo/game/game.go]:
```suggestion
// Shift all values from xEnd to xStart in one row
```

File: advisor-go/src/quixo/scorer/scorer.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+import (
+	"quixo/game"
+)
+
+/*
+ * GetGridScore return the score of the board
+ * the score of the board is the max score of all straight lines or diagonal
+ */
+func GetBoardScore(board game.Board) int {
```
Comment (fzaninotto) [inline advisor-go/src/quixo/scorer/scorer.go]: I don't understand that it doesn't depend on the player

File: advisor-go/src/quixo/scorer/scorer.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+func getMaxLinesScore(board game.Board) int {
+	size := len(board.Grid)
+	scores := []int{}
+	for i := 0; i < size; i++ {
+		scores = append(scores, getLineScore(board.Grid[i], board.Player))
+	}
+	return getMaxInArray(scores)
+}
+
+func getLineScore(line []int, player int) int {
```
Comment (fzaninotto) [inline advisor-go/src/quixo/scorer/scorer.go]: player shouldn't be an int IMO, but an enum

File: advisor-go/src/quixo/scorer/scorer.go (hunk 4 above)
Comment (HALLERPierre) [reply advisor-go/src/quixo/scorer/scorer.go]: It's the score for the player. If 4 cubes belonging to the player are aligned, it'll return 4. The cubes of the other player don't count.

File: advisor-go/src/quixo/game/game.go [hunk 6]
```diff
@@ -111,7 +111,7 @@ func isPlayerCube(cubeValue int, playerValue int) bool {
 	return cubeValue == 0 || cubeValue == playerValue
 }
 
-// Shift all value from xEnd to xStart in one row
+// Shift all values from xEnd to xStart in one row
```
Comment (alexisjanvier) [inline advisor-go/src/quixo/game/game.go]: Don't hesitate to do git `squash` or git `fixup` for this kind of fix.

On my side, I made two shortcuts for these cases: `git fixup` and `git please`

```
// in ~/.gitconfig
[alias]
    commend = commit --amend --no-edit
    fixup = "!git add -p && git commend"
    please = push --force-with-lease
```

Comment (HALLERPierre) [reply advisor-go/src/quixo/game/game.go]: Thank you for the tip :+1:

File: advisor-go/src/quixo/server/server.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
+package server
+
+import (
+	"net/http"
+	"strconv"
+)
+
+const port int = 8001
```
Comment (alexisjanvier) [inline advisor-go/src/quixo/server/server.go]: `port` could be declared as an environment variable (to be injected via the `docker-composer.yml`)

File: advisor-go/src/quixo/game/game.go (hunk 6 above)
Comment (zyhou) [reply advisor-go/src/quixo/game/game.go]: If you use oh-my-zsh, you have a lot of shortcut, https://github.com/robbyrussell/oh-my-zsh/tree/master/plugins/git

I always use `gc!` (git commit -v --amend) in your case.

File: advisor-go/src/quixo/simulation/simulation.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
+package simulation
```
Comment (alexisjanvier) [inline advisor-go/src/quixo/simulation/simulation.go]: `simulation` => `advice` ?

File: app/src/Manager/AdvisorManager.php [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+     * @return int
+     */
+    public function getAdvice(Game $game)
+    {
+        $body = [
+            'Grid' => $game->getBoard(),
+            'Player' => $game->getCurrentPlayer(),
+        ];
+        $response = $this->post(json_encode($body));
+        dump($response);
```
Comment (alexisjanvier) [inline app/src/Manager/AdvisorManager.php]: debug spotted

File: app/src/Manager/AdvisorManager.php [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+use App\Entity\Game;
+use App\Repository\GameRepository;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\HttpFoundation\Response;
+use Unirest\Request as Request;
+
+class AdvisorManager
+{
+    private const ADVISOR_URL = 'http://advisor:8001';
```
Comment (alexisjanvier) [inline app/src/Manager/AdvisorManager.php]: It should be in configuration (possibly with the port from an environment variable)

File: advisor-go/src/quixo/scorer/scorer.go (hunk 5 above)
Comment (HALLERPierre) [reply advisor-go/src/quixo/scorer/scorer.go]: I agree it's not good. But enum doesn't exist in Golang (or with a specific struct) and the advisor won't have to manipulate the player other than for comparison. So I'll leave it like that to save time.

File: advisor-go/src/quixo/simulation/simulation.go (hunk 8 above)
Comment (HALLERPierre) [reply advisor-go/src/quixo/simulation/simulation.go]: I like simulation because it's understandable that the package will test a lot of possibility. Maybe I should split that in two package in the future

File: advisor-go/src/quixo/game/game.go (hunk 1 above)
Comment (HALLERPierre) [reply advisor-go/src/quixo/game/game.go]: Added commentary and lambda to make it more clearer

File: app/src/Manager/AdvisorManager.php [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    private function post($body): Response
+    {
+        $headers = ['Accept' => 'application/json'];
+        return Request::post($this->adivsorUrl.'/best-move', $headers, $body);
+    }
+
+    /**
+     * Return the team of the player stored in the sessoin
```
Comment (alexisjanvier) [inline app/src/Manager/AdvisorManager.php]: `the team of the player` ? I don't understand ...

File: app/.env [hunk 12]
```diff
@@ -17,3 +17,4 @@
 APP_ENV=dev
 APP_SECRET=fb33e31afe982a62ee0105780eed23a5
```
Comment (alexisjanvier) [inline app/.env]: a secret should not be visible in the repo

**Comment by alexisjanvier** [comment]
Hey, It works on my laptop :) Good job :+1:

File: advisor-go/src/quixo/game/game.go [hunk 13]
```diff
@@ -1,2 +1,183 @@
...
+	isOnEnd := func(index int, size int) bool {
+		return index == size-1
+	}
+	isOnEdge := func(index int, size int) bool {
+		return isOnStart(index) || isOnEnd(index, size)
+	}
+	indexStart := 0
+	indexEnd := size - 1
+
+	// If on the top or the bottom of the board, else it's on the left or right
```
[the remark this answers is not in the document]
Comment (HALLERPierre) [reply advisor-go/src/quixo/game/game.go]: Yes it could ! I need the size parameter to know if it's on the edge of the board !

File: app/src/Manager/AdvisorManager.php (hunk 11 above)
Comment (HALLERPierre) [reply app/src/Manager/AdvisorManager.php]: Yes I don't understand neither... Fix in next PR
